File: src/aish/memory/manager.py
from __future__ import annotations

import datetime as dt
import re
from pathlib import Path

from aish.memory.config import MemoryConfig
from aish.memory.models import MemoryCategory, MemoryEntry


class MemoryManager:
    """Single-file Markdown memory manager backed by MEMORY.md."""

    _SECTION_HEADERS = {
        MemoryCategory.PREFERENCE: "## Preferences",
        MemoryCategory.ENVIRONMENT: "## Environment",
        MemoryCategory.SOLUTION: "## Solutions",
        MemoryCategory.PATTERN: "## Patterns",
        MemoryCategory.OTHER: "## Other",
    }
    _ENTRY_RE = re.compile(
        r"^- \[#(?P<id>\d+)\] (?P<content>.*?)(?: <!-- (?P<meta>.*?) -->)?$"
    )
    _MANUAL_ENTRY_RE = re.compile(r"^- (?P<content>.+)$")
    _META_RE = re.compile(r"(?P<key>[a-z_]+)=(?P<value>[^ ]+)")
    _WORD_RE = re.compile(r"[a-z0-9_./-]+")

    def __init__(self, config: MemoryConfig):
        self.config = config
        self.memory_dir = Path(config.data_dir).expanduser().resolve()
        self.memory_dir.mkdir(parents=True, exist_ok=True)
        self.memory_md = self.memory_dir / "MEMORY.md"
        self._ensure_memory_md()

    def close(self) -> None:
        """Release resources."""

    def store(
        self,
        content: str,
        category: MemoryCategory,
        source: str = "explicit",
        tags: str = "",
        importance: float = 0.5,
    ) -> int:
        _ = tags
        normalized_content = self._normalize_content(content)
        if not normalized_content:
            raise ValueError("Memory content cannot be empty")

        lines = self._read_lines()
        entries = self._parse_entries(lines)
        existing = self._find_duplicate(entries, category, normalized_content)
        if existing is not None:
            return existing.id

        entry_id = max((entry.id for entry in entries if entry.id > 0), default=0) + 1
        created_at = dt.datetime.utcnow().replace(microsecond=0).isoformat()
        entry = MemoryEntry(
            id=entry_id,
            source=source,
            category=category,
            content=normalized_content,
            importance=importance,
            created_at=created_at,
        )
        entry_line = self._format_entry_line(entry)
        updated_lines = self._insert_entry_line(lines, category, entry_line)
        self._write_lines(updated_lines)
        return entry_id

    def recall(self, query: str, limit: int = 5) -> list[MemoryEntry]:
        normalized_query = query.strip().lower()
        if not normalized_query:
            return []

        query_tokens = set(self._tokenize(normalized_query))
        if not query_tokens and len(normalized_query) < 2:
            return []

        scored: list[tuple[int, float, int, MemoryEntry]] = []
        for entry in self._parse_entries(self._read_lines()):
            score = self._score_entry(entry, normalized_query, query_tokens)
            if score <= 0:
                continue
            scored.append((score, entry.importance, entry.id, entry))

        scored.sort(key=lambda item: (item[0], item[1], item[2]), reverse=True)
        return [item[3] for item in scored[:limit]]

    def delete(self, entry_id: int) -> None:
        lines = self._read_lines()
        updated_lines = [
            line for line in lines if not line.startswith(f"- [#{entry_id}] ")
        ]
        self._write_lines(updated_lines)

    def list_recent(self, limit: int = 10) -> list[MemoryEntry]:
        entries = [
            entry for entry in self._parse_entries(self._read_lines()) if entry.id > 0
        ]
        entries.sort(key=lambda entry: entry.id, reverse=True)
        return entries[:limit]

    def get_session_context(self) -> str:
        text = self.memory_md.read_text().strip()
        return text if self._has_stored_entries(text) else ""

    def get_system_prompt_section(self) -> str:
        return (
            "## Memory System\n"
            "You have persistent long-term memory stored in MEMORY.md.\n"
            "1. Before relying on prior preferences, environment details, or project decisions, use the memory tool with action search.\n"
            "2. When the user shares an important durable fact, use the memory tool with action store.\n"
            "3. Keep stored memories short, factual, and reusable. Avoid saving transient chatter.\n"
            f"4. The memory file lives in {self.memory_md}.\n"
        )

    def _ensure_memory_md(self) -> None:
        if self.memory_md.exists():
            return
        self.memory_md.write_text(
            "# Long-term Memory\n\n"
            "Persistent facts about the user, environment, and project.\n\n"
            "## Preferences\n\n"
            "## Environment\n\n"
            "## Solutions\n\n"
            "## Patterns\n\n"
            "## Other\n"
        )

    def _read_lines(self) -> list[str]:
        return self.memory_md.read_text().splitlines()

    def _write_lines(self, lines: list[str]) -> None:
        self.memory_md.write_text("\n".join(lines).rstrip() + "\n")

    def _parse_entries(self, lines: list[str]) -> list[MemoryEntry]:
        entries: list[MemoryEntry] = []
        current_category: MemoryCategory | None = None
        for line in lines:
            category = self._category_for_header(line.strip())
            if category is not None:
                current_category = category
                continue
            if current_category is None:
                continue

            explicit_match = self._ENTRY_RE.match(line)
            if explicit_match:
                meta = self._parse_meta(explicit_match.group("meta") or "")
                entries.append(
                    MemoryEntry(
                        id=int(explicit_match.group("id")),
                        source=meta.get("source", "manual"),
                        category=current_category,
                        content=explicit_match.group("content").strip(),
                        importance=float(meta.get("importance", "0.5")),
                        created_at=meta.get("created_at"),
                    )
                )
                continue

            manual_match = self._MANUAL_ENTRY_RE.match(line)
            if manual_match:
                entries.append(
                    MemoryEntry(
                        id=0,
                        source="manual",
                        category=current_category,
                        content=manual_match.group("content").strip(),
                        importance=0.5,
                    )
                )
        return entries

    def _parse_meta(self, meta_text: str) -> dict[str, str]:
        return {
            match.group("key"): match.group("value")
            for match in self._META_RE.finditer(meta_text)
        }

    def _find_duplicate(
        self,
        entries: list[MemoryEntry],
        category: MemoryCategory,
        content: str,
    ) -> MemoryEntry | None:
        normalized = content.casefold()
        for entry in entries:
            if entry.category == category and entry.content.casefold() == normalized:
                return entry
        return None

    def _insert_entry_line(
        self, lines: list[str], category: MemoryCategory, entry_line: str
    ) -> list[str]:
        header = self._SECTION_HEADERS[category]
        updated_lines = list(lines)
        header_index = next(
            (index for index, line in enumerate(updated_lines) if line.strip() == header),
            None,
        )
        if header_index is None:
            if updated_lines and updated_lines[-1] != "":
                updated_lines.append("")
            updated_lines.extend([header, "", entry_line])
            return updated_lines

        insert_index = header_index + 1
        while (
            insert_index < len(updated_lines)
            and not updated_lines[insert_index].startswith("## ")
        ):
            insert_index += 1

        while insert_index > header_index + 1 and updated_lines[insert_index - 1] == "":
            insert_index -= 1

        updated_lines.insert(insert_index, entry_line)
        if insert_index + 1 >= len(updated_lines) or updated_lines[insert_index + 1] != "":
            updated_lines.insert(insert_index + 1, "")
        return updated_lines

    def _format_entry_line(self, entry: MemoryEntry) -> str:
        meta = (
            f"source={self._sanitize_meta_value(entry.source)} "
            f"importance={entry.importance:.2f} "
            f"created_at={self._sanitize_meta_value(entry.created_at or '')}"
        )
        return f"- [#{entry.id}] {entry.content} <!-- {meta.strip()} -->"

    def _score_entry(
        self,
        entry: MemoryEntry,
        normalized_query: str,
        query_tokens: set[str],
    ) -> int:
        content = entry.content.casefold()
        content_tokens = set(self._tokenize(content))
        overlap = len(query_tokens & content_tokens)
        if normalized_query in content:
            overlap += 5
        if any(token in content for token in query_tokens):
            overlap += 2
        return overlap

    def _tokenize(self, text: str) -> list[str]:
        return self._WORD_RE.findall(text.casefold())

    def _category_for_header(self, line: str) -> MemoryCategory | None:
        for category, header in self._SECTION_HEADERS.items():
            if line == header:
                return category
        return None

    def _has_stored_entries(self, text: str) -> bool:
        return "- [#" in text or bool(re.search(r"^## .*\n\n- ", text, flags=re.MULTILINE))

    def _normalize_content(self, content: str) -> str:
        return " ".join(content.strip().split())

    def _sanitize_meta_value(self, value: str) -> str:
        return value.replace(" ", "_")

File: crates/aish-memory/src/manager.rs
use crate::models::MemoryEntry;
use aish_core::{AishError, MemoryCategory};
use chrono::Utc;
use std::io::Write;
use std::path::{Path, PathBuf};

const HEADER: &str = "# Memory\n";

/// Memory manager backed by a single MEMORY.md file.
pub struct MemoryManager {
    memory_file: PathBuf,
    entries: Vec<MemoryEntry>,
    next_id: i64,
}

impl MemoryManager {
    /// Create or open a memory file.
    pub fn new(memory_file: PathBuf) -> aish_core::Result<Self> {
        if !memory_file.exists() {
            // Create parent directories and the file with just the header
            if let Some(parent) = memory_file.parent() {
                std::fs::create_dir_all(parent).map_err(|e| {
                    AishError::Memory(format!("cannot create directory {:?}: {}", parent, e))
                })?;
            }
            let mut f = std::fs::File::create(&memory_file)
                .map_err(|e| AishError::Memory(format!("cannot create memory file: {}", e)))?;
            f.write_all(HEADER.as_bytes())
                .map_err(|e| AishError::Memory(format!("cannot write memory header: {}", e)))?;
            return Ok(Self {
                memory_file,
                entries: Vec::new(),
                next_id: 1,
            });
        }

        let entries = parse_file(&memory_file)?;
        let next_id = entries.iter().map(|e| e.id).max().unwrap_or(0) + 1;

        Ok(Self {
            memory_file,
            entries,
            next_id,
        })
    }

    /// Return the default memory file path: `~/.config/aish/MEMORY.md`.
    pub fn default_path() -> PathBuf {
        dirs::config_dir()
            .unwrap_or_else(|| PathBuf::from("."))
            .join("aish")
            .join("MEMORY.md")
    }

    /// Store a new memory entry. Returns the assigned ID.
    /// If an entry with the same content and category already exists, returns
    /// the existing ID without creating a duplicate.
    pub fn store(
        &mut self,
        content: &str,
        category: MemoryCategory,
        source: &str,
        importance: f64,
    ) -> aish_core::Result<i64> {
        let content_trimmed = content.trim();

        // Duplicate detection: same content (case-insensitive) + same category
        let content_lower = content_trimmed.to_lowercase();
        for entry in &self.entries {
            if entry.category == category && entry.content.to_lowercase() == content_lower {
                return Ok(entry.id);
            }
        }

        let now = Utc::now().format("%Y-%m-%d").to_string();
        let id = self.next_id;
        self.next_id += 1;

        let entry = MemoryEntry {
            id,
            source: source.to_string(),
            category,
            content: content_trimmed.to_string(),
            importance,
            tags: String::new(),
            created_at: Some(now.clone()),
            last_accessed_at: Some(now),
            access_count: 0,
        };

        self.entries.push(entry);
        self.persist()?;
        Ok(id)
    }

    /// Recall memories matching a query, sorted by relevance.
    ///
    /// Relevance = (number of matching query words) * importance.
    /// Access stats are updated for matched entries.
    pub fn recall(&mut self, query: &str, limit: usize) -> Vec<&MemoryEntry> {
        let query_words: Vec<String> = query.split_whitespace().map(|w| w.to_lowercase()).collect();

        if query_words.is_empty() {
            // Return entries sorted by importance when query is empty
            let mut indices: Vec<usize> = (0..self.entries.len()).collect();
            indices.sort_by(|a, b| {
                self.entries[*b]
                    .importance
                    .partial_cmp(&self.entries[*a].importance)
                    .unwrap_or(std::cmp::Ordering::Equal)
            });
            return indices
                .into_iter()
                .take(limit)
                .map(|i| &self.entries[i])
                .collect();
        }

        let now = Utc::now().format("%Y-%m-%d").to_string();

        // Compute relevance scores
        let mut scored: Vec<(usize, f64)> = Vec::new();
        for (idx, entry) in self.entries.iter().enumerate() {
            let content_lower = entry.content.to_lowercase();
            let match_count = query_words
                .iter()
                .filter(|w| content_lower.contains(w.as_str()))
                .count() as f64;
            if match_count > 0.0 {
                let score = match_count * entry.importance;
                scored.push((idx, score));
            }
        }

        // Sort by score descending, then by importance as tie-breaker
        scored.sort_by(|a, b| {
            b.1.partial_cmp(&a.1)
                .unwrap_or(std::cmp::Ordering::Equal)
                .then_with(|| {
                    self.entries[b.0]
                        .importance
                        .partial_cmp(&self.entries[a.0].importance)
                        .unwrap_or(std::cmp::Ordering::Equal)
                })
        });

        // Update access stats for matched entries
        for (idx, _) in &scored {
            self.entries[*idx].access_count += 1;
            self.entries[*idx].last_accessed_at = Some(now.clone());
        }

        // Persist updated access stats (best-effort)
        let _ = self.persist();

        scored
            .into_iter()
            .take(limit)
            .map(|(idx, _)| &self.entries[idx])
            .collect()
    }

    /// Remove a memory entry by ID. Returns true if found and removed.
    pub fn remove(&mut self, id: i64) -> aish_core::Result<bool> {
        let before = self.entries.len();
        self.entries.retain(|e| e.id != id);
        let removed = self.entries.len() < before;
        if removed {
            self.persist()?;
        }
        Ok(removed)
    }

    /// List all stored memory entries.
    pub fn list(&self) -> &[MemoryEntry] {
        &self.entries
    }

    /// Generate a system prompt section describing the memory system.
    /// This should be appended to the LLM system prompt when memory is enabled.
    pub fn get_system_prompt_section(&self) -> String {
        format!(
            "## Memory System\n\
             You have persistent long-term memory stored in MEMORY.md.\n\
             1. Before relying on prior preferences, environment details, or project decisions, use the memory tool with action search.\n\
             2. When the user shares an important durable fact, use the memory tool with action store.\n\
             3. Keep stored memories short, factual, and reusable. Avoid saving transient chatter.\n\
             4. The memory file lives in {}.\n",
            self.memory_file.display()
        )
    }

    /// Get the full memory file content for session context injection.
    /// Returns an empty string if no entries exist.
    pub fn get_session_context(&self) -> String {
        if self.entries.is_empty() {
            return String::new();
        }
        let mut out = String::from(HEADER);
        for entry in &self.entries {
            let category = format_category(&entry.category);
            let date = entry.created_at.as_deref().unwrap_or("unknown");
            out.push_str(&format!(
                "\n## [{}] [{}] Source: {} | {}\n{}\n",
                entry.id, category, entry.source, date, entry.content,
            ));
        }
        out
    }

    /// Persist the full entry list to the MEMORY.md file.
    fn persist(&self) -> aish_core::Result<()> {
        let mut out = String::from(HEADER);

        for entry in &self.entries {
            let category = format_category(&entry.category);
            let date = entry.created_at.as_deref().unwrap_or("unknown");
            out.push_str(&format!(
                "\n## [{}] [{}] Source: {} | {}\n{}\n",
                entry.id, category, entry.source, date, entry.content,
            ));
        }

        std::fs::write(&self.memory_file, out)
            .map_err(|e| AishError::Memory(format!("failed to write memory file: {}", e)))
    }
}

// ---------------------------------------------------------------------------
// File format parsing
// ---------------------------------------------------------------------------

/// Parse an existing MEMORY.md file into a list of entries.
fn parse_file(path: &Path) -> aish_core::Result<Vec<MemoryEntry>> {
    let content = std::fs::read_to_string(path)
        .map_err(|e| AishError::Memory(format!("cannot read memory file {:?}: {}", path, e)))?;

    let mut entries = Vec::new();
    let mut current_lines: Vec<String> = Vec::new();
    let mut current_meta: Option<ParsedMeta> = None;

    for line in content.lines() {
        if let Some(rest) = line.strip_prefix("## [") {
            // Flush previous entry
            if let Some(meta) = current_meta.take() {
                let body = current_lines.join("\n").trim().to_string();
                entries.push(build_entry(meta, body));
                current_lines.clear();
            }
            // Parse new header
            if let Some(meta) = parse_header(rest) {
                current_meta = Some(meta);
            }
        } else if current_meta.is_some() {
            current_lines.push(line.to_string());
        }
    }

    // Flush last entry
    if let Some(meta) = current_meta.take() {
        let body = current_lines.join("\n").trim().to_string();
        entries.push(build_entry(meta, body));
    }

    Ok(entries)
}

struct ParsedMeta {
    id: i64,
    category: MemoryCategory,
    source: String,
    date: Option<String>,
}

/// Parse a header line after the leading `## [` has been stripped.
///
/// Expected format: `id] [Category] Source: source | date`
fn parse_header(rest: &str) -> Option<ParsedMeta> {
    // Find `]` to get the id
    let bracket_pos = rest.find(']')?;
    let id: i64 = rest[..bracket_pos].trim().parse().ok()?;
    let rest = &rest[bracket_pos + 1..];

    // Find `[Category]`
    let rest = rest.trim_start();
    let rest = rest.strip_prefix('[')?;
    let bracket2 = rest.find(']')?;
    let category_str = &rest[..bracket2];
    let category = parse_category(category_str)?;
    let rest = &rest[bracket2 + 1..];

    // Find "Source: ..."
    let rest = rest.trim_start();
    let rest = rest.strip_prefix("Source:")?;
    let rest = rest.trim_start();

    // Split on "|"
    let mut parts = rest.splitn(2, '|');
    let source = parts.next().unwrap_or("").trim().to_string();
    let date = parts.next().map(|d| d.trim().to_string());

    Some(ParsedMeta {
        id,
        category,
        source,
        date,
    })
}

fn build_entry(meta: ParsedMeta, content: String) -> MemoryEntry {
    MemoryEntry {
        id: meta.id,
        source: meta.source,
        category: meta.category,
        content,
        importance: 1.0,
        tags: String::new(),
        created_at: meta.date,
        last_accessed_at: None,
        access_count: 0,
    }
}

fn format_category(cat: &MemoryCategory) -> &'static str {
    match cat {
        MemoryCategory::Preference => "Preference",
        MemoryCategory::Environment => "Environment",
        MemoryCategory::Solution => "Solution",
        MemoryCategory::Pattern => "Pattern",
        MemoryCategory::Other => "Other",
    }
}

fn parse_category(s: &str) -> Option<MemoryCategory> {
    match s.trim() {
        "Preference" => Some(MemoryCategory::Preference),
        "Environment" => Some(MemoryCategory::Environment),
        "Solution" => Some(MemoryCategory::Solution),
        "Pattern" => Some(MemoryCategory::Pattern),
        "Other" => Some(MemoryCategory::Other),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_roundtrip() {
        let dir = std::env::temp_dir().join("aish_memory_test_roundtrip");
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("MEMORY.md");

        let mut mgr = MemoryManager::new(path.clone()).unwrap();

        let id1 = mgr
            .store(
                "I prefer dark theme",
                MemoryCategory::Preference,
                "auto",
                1.0,
            )
            .unwrap();
        let id2 = mgr
            .store(
                "db port is 5432",
                MemoryCategory::Environment,
                "manual",
                0.8,
            )
            .unwrap();

        assert_eq!(id1, 1);
        assert_eq!(id2, 2);
        assert_eq!(mgr.list().len(), 2);

        // Reload from disk
        let mgr2 = MemoryManager::new(path.clone()).unwrap();
        assert_eq!(mgr2.list().len(), 2);
        assert_eq!(mgr2.list()[0].content, "I prefer dark theme");
        assert_eq!(mgr2.list()[1].content, "db port is 5432");

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_recall() {
        let dir = std::env::temp_dir().join("aish_memory_test_recall");
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("MEMORY.md");

        let mut mgr = MemoryManager::new(path).unwrap();
        mgr.store(
            "dark theme preference",
            MemoryCategory::Preference,
            "auto",
            1.0,
        )
        .unwrap();
        mgr.store(
            "database port 5432",
            MemoryCategory::Environment,
            "manual",
            0.8,
        )
        .unwrap();
        mgr.store(
            "use rust for systems code",
            MemoryCategory::Pattern,
            "auto",
            0.6,
        )
        .unwrap();

        let results = mgr.recall("dark theme", 10);
        assert_eq!(results.len(), 1);
        assert!(results[0].content.contains("dark theme"));

        let results = mgr.recall("port", 10);
        assert_eq!(results.len(), 1);
        assert!(results[0].content.contains("port 5432"));

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_remove() {
        let dir = std::env::temp_dir().join("aish_memory_test_remove");
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("MEMORY.md");

        let mut mgr = MemoryManager::new(path).unwrap();
        mgr.store("entry one", MemoryCategory::Other, "test", 1.0)
            .unwrap();
        let id = mgr
            .store("entry two", MemoryCategory::Other, "test", 1.0)
            .unwrap();
        assert_eq!(mgr.list().len(), 2);

        let removed = mgr.remove(id).unwrap();
        assert!(removed);
        assert_eq!(mgr.list().len(), 1);
        assert_eq!(mgr.list()[0].content, "entry one");

        // Removing non-existent returns false
        let removed = mgr.remove(999).unwrap();
        assert!(!removed);

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_parse_header() {
        let meta = parse_header("1] [Preference] Source: auto | 2024-01-01").unwrap();
        assert_eq!(meta.id, 1);
        assert_eq!(meta.source, "auto");
        assert_eq!(meta.date.as_deref(), Some("2024-01-01"));
    }

    #[test]
    fn test_default_path() {
        let path = MemoryManager::default_path();
        assert!(path.to_string_lossy().contains("aish"));
        assert!(path.to_string_lossy().contains("MEMORY.md"));
    }

    #[test]
    fn test_duplicate_detection() {
        let dir = std::env::temp_dir().join("aish_memory_test_dup");
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("MEMORY.md");

        let mut mgr = MemoryManager::new(path).unwrap();
        let id1 = mgr
            .store(
                "I prefer dark theme",
                MemoryCategory::Preference,
                "auto",
                1.0,
            )
            .unwrap();
        let id2 = mgr
            .store(
                "I prefer dark theme",
                MemoryCategory::Preference,
                "auto",
                1.0,
            )
            .unwrap();
        // Duplicate should return same ID
        assert_eq!(id1, id2);
        assert_eq!(mgr.list().len(), 1);

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_duplicate_case_insensitive() {
        let dir = std::env::temp_dir().join("aish_memory_test_dup_case");
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("MEMORY.md");

        let mut mgr = MemoryManager::new(path).unwrap();
        let id1 = mgr
            .store(
                "Database port 5432",
                MemoryCategory::Environment,
                "auto",
                1.0,
            )
            .unwrap();
        let id2 = mgr
            .store(
                "database PORT 5432",
                MemoryCategory::Environment,
                "auto",
                1.0,
            )
            .unwrap();
        assert_eq!(id1, id2);
        assert_eq!(mgr.list().len(), 1);

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_duplicate_different_category_allowed() {
        let dir = std::env::temp_dir().join("aish_memory_test_dup_cat");
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("MEMORY.md");

        let mut mgr = MemoryManager::new(path).unwrap();
        let id1 = mgr
            .store("same content", MemoryCategory::Preference, "auto", 1.0)
            .unwrap();
        let id2 = mgr
            .store("same content", MemoryCategory::Solution, "auto", 1.0)
            .unwrap();
        // Different category = different entry
        assert_ne!(id1, id2);
        assert_eq!(mgr.list().len(), 2);

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_system_prompt_section() {
        let dir = std::env::temp_dir().join("aish_memory_test_sysprompt");
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("MEMORY.md");

        let mgr = MemoryManager::new(path).unwrap();
        let section = mgr.get_system_prompt_section();
        assert!(section.contains("Memory System"));
        assert!(section.contains("MEMORY.md"));

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_session_context_empty() {
        let dir = std::env::temp_dir().join("aish_memory_test_ctx_empty");
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("MEMORY.md");

        let mgr = MemoryManager::new(path).unwrap();
        let ctx = mgr.get_session_context();
        assert!(ctx.is_empty());

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_session_context_with_entries() {
        let dir = std::env::temp_dir().join("aish_memory_test_ctx");
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("MEMORY.md");

        let mut mgr = MemoryManager::new(path).unwrap();
        mgr.store("test entry", MemoryCategory::Other, "test", 1.0)
            .unwrap();
        let ctx = mgr.get_session_context();
        assert!(!ctx.is_empty());
        assert!(ctx.contains("test entry"));

        let _ = std::fs::remove_dir_all(&dir);
    }
}
